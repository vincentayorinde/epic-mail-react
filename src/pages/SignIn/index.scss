@import url('https://fonts.googleapis.com/css?family=Montserrat');

@media (min-width: 200px) and (max-width: 850px) {
  .container-1 .box-1{
    display: none;
    flex-direction: row;
  }
  .container-1 .box-3{
    display: none;
    align-items: stretch;
  }
}

* {
  margin: 0;
  padding: 0;
}

html, body, .viewport{
  background-color: #f4f7f9;
  font-family: 'Montserrat', sans-serif;
  min-height: 100%;
 
}

.container-1{
  display: flex;
  height: 100vh;
}
.container-1 div{
  flex-direction: column;
}

.box-1{
  flex: 6;
  background-image: url('../../assets/images/bg-home.jpg');
  background-size: cover;
  background-repeat: no-repeat;
  background-position: right bottom;

}
.box-1 .content{
  font-weight: bolder;
 padding:20px 20px 20px 20px;
 flex: 1;
 background-color:rgba(0, 74, 212, 0.5);
 filter: alpha(opacity=40);
 height: 94vh;
 display: flex;
}
.box-1 .content p{
  color: #fff;
  opacity: unset;
  font-size: 35px;
  margin: 10px 0px 10px 40px;

}
.box-1 .content br{
  margin-bottom: 10px
}
.box-1 .content .links{
  align-self: flex-start;
  justify-content: center;
  margin-top: auto !important;
}
.box-1 .content .links ul{
  display: inline-block;
  list-style: none;
  margin-bottom: auto !important;
  margin-top: auto !important;
}
.box-1 .content .links ul li{
  display: inline-block;
  list-style: none;
  margin-right: 30px;
  margin-left: 40px;
  margin-bottom: 40px;
}
.box-1 .content .links ul li a{
  text-decoration: none;
  color: #fff;
}
.box-2{
  flex: 4;
  padding:20px 20px 20px 20px;
  box-shadow: 1px -20px 30px 1px #796969;
  background-color: #fff;
}

.box-2 .signup h2{
  margin-bottom: 40px;
  color: #38d85a;
  font-weight: bolder;
}

.box-2 .signup p{
  color: #bcc0db;
  margin-bottom: 10px;
  margin-top: 15px;
  text-transform: uppercase;
}
.box-2 .signup a{
  text-decoration: none;
  color: #38d85a;
  font-weight: bold;
}

.box-2
input[type='email'],
input[type='text'],
input[type='password']{
  border: 0px;
  border-bottom: 1px solid rgba(0, 74, 212, 0.3);
  background-color: #fff;
  padding: 10px;
  width: 90%;
}


.box-2 
input[type='submit']{
  margin-top: 20px;
  background-color: #38d85a;
  color: #fff;
  cursor: pointer;
  border: 0px;
  font-family: 'Montserrat', sans-serif;
  padding: 15px;
  box-shadow: -10px 1px 30px 1px #ccc;
  width: 200px;
  border-radius: 30px;
}
.box-2 
input[type='submit']:hover {
  margin-top: 20px;
  background-color: #208636;
  color: #fff;
  cursor: pointer;
  border: 0px;
  font-family: 'Montserrat', sans-serif;
  padding: 15px;
  box-shadow: -10px 1px 30px 1px #ccc;
  transition-duration: 0.55s;
}
.box-3{
  flex: 2;
  background-color:rgba(0, 74, 212, 0.5);
  filter: alpha(opacity=40);
  min-height: 100%;
}
.box-3 .content{
  width: 200px;
}