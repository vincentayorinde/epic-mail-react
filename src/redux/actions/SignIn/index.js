import { axiosRequest, populateLocalStorage } from '../../../utils';

export const cleanSignIn = () => ({ type: 'CLEAN_SIGNIN' });
export const signInPending = () => ({ type: 'SIGNIN_PENDING', payload: { isCompleted: false, isLoading: true, error: null } });
export const signInSuccess = user => ({ type: 'SIGNIN_SUCCESS', payload: { user, isCompleted: true, isAuthenticated: true } });
export const signInError = payload => ({ type: 'SIGNIN_ERROR', payload });
export const signInUser = () => ({ type: 'SIGNIN_USER', payload: { isSubmit: true } });
export const setSignedInUser = payload => ({ type: 'SET_SIGNED_USER', payload });

export const signInAction = (signInData, history) => async (dispatch) => {
  dispatch(signInPending());
  dispatch(cleanSignIn());
  // dispatch(signInUser());
  try {
    const result = await axiosRequest({ path: '/api/v2/auth/login', payload: signInData, method: 'post' }); 
    populateLocalStorage(result);
    dispatch(signInSuccess(result));
  } catch (e) {
    /* istanbul ignore next */
    const { response } = e;
    /* istanbul ignore next */
    const message = response.data.error || response;
    /* istanbul ignore next */
    dispatch(signInError(message));
  }
};

export const getUser = () => (dispatch) => {
  /* istanbul ignore next */
  if (localStorage.token && localStorage.user) {
    let user = localStorage.getItem('user');
    user = JSON.parse(user);
    dispatch(setSignedInUser(user));
  }
};
