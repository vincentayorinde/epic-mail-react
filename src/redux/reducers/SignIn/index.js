const initialState = {
  isLoading: false,
  error: null,
  user: {},
  isAuthenticated: false,
  isCompleted: false,
  isSubmit: false,
};

const signInReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SIGNIN_PENDING':
      return {
        ...state,
        ...action.payload,
      };
    case 'SIGNIN_SUCCESS':
      return {
        ...state,
        ...action.payload,
        isSubmit: false,
      };
    case 'SIGNIN_ERROR':
      return {
        ...state,
        error: action.payload,
        isSubmit: false,
      };
    case 'CLEAN_SIGNIN':
      return {
        ...state,
        isCompleted: true,
        error: null,
      };
    case 'SET_SIGNED_USER':
      return {
        ...state,
        isAuthenticated: true,
        user: { ...action.payload },
      };
    case 'SIGNIN_USER':
      return {
        ...state,
        isAuthenticated: true,
        isCompleted: true,
        ...action.payload,
      };
    default:
      return state;
  }
};

export default signInReducer;
