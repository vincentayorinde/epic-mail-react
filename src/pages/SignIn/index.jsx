import './index.scss';
import React, { Fragment, useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { toast } from 'react-toastify';
import { signInAction, cleanSignIn } from '../../redux/actions/SignIn';
import '../../assets/css/react-toastify.scss';
import Input from '../../components/Input';
import Button from '../../components/Button';


const SignIn = (props) => {
  const {
    isCompleted, history, isSubmit, error
  } = props;
  const [values, setValues] = useState({
    email: '',
    password: '',
  });
  const onChange = (e) => {
    e.persist();
    setValues(prevState => ({ ...prevState, [e.target.name]: e.target.value }));
  };

  const submit = (e) => {
    e.preventDefault();
    /* istanbul ignore next */
    if (!values.email && !values.password === true) {
      /* istanbul ignore next */
      toast.error(<h4 className="text-center">All fields are required</h4>);
      /* istanbul ignore next */
      return;
    }
    props.onSignIn(values);
  };

  useEffect(() => {
    if (isCompleted) {
      toast.success(<h4 id="toast-success" className="text-center">Login successful</h4>);
      /* istanbul ignore next */
      history && history.push('/inbox');
    }
    if (error) {
      if (Array.isArray(error)) {
        for (let i = 0; i < error.length; i += 1) {
          toast.error(<h4 className="text-center">{error[i].message}</h4>);
        }
      } else {
        toast.error(<h4 className="text-center">{error}</h4>);
      }
    }
    /* istanbul ignore next */
    return () => {
      /* istanbul ignore next */
      props.cleanUp();
    };
  }, [isCompleted, error]);
  return (
    <Fragment>
    <div className="container-1">
    <div className="box-1">
      <div className="content">
          <p>Experience</p>
          <p>the new way to</p>
          <p> send messages</p>
          <p>Use EPIC Mail <em>today!</em></p>
          <div className="links">
              <ul>
                <li><a href="#">Terms</a></li>
                <li><a href="#">Privacy</a></li>
                <li><a href="#">APIs</a></li>
              </ul>
            </div>
      </div>
      
    </div>
    <div className="box-2">
      <div className="signup">
        <p>Welcome to EPIC Mail</p>
          <h2>Sign In</h2>
            <form onSubmit={submit}>
              <p>Email</p>
              <Input 
              inputType="email"
              name="email"
              id="email"
              onChange={onChange}
              pattern="^[\w.]+@[\w]{2,20}.[a-z]{2,10}$"
              title="user@email.com" 
              required />
              <p>Password</p>
              <Input type="password" onChange={onChange} name="password" id="password" required />
              <Button
                isSubmit={isSubmit} 
                type="submit" 
                name="Sign In" 
              />
            </form> 
            <br /><br />
            <i className="fas fa-chevron-right"></i> Don't have an account? <Link to="/">Sign Up</Link>
      </div>
    </div>
    <div className="box-3">
        <div className="content">
        </div>
    </div>
    </div>
  </Fragment>
  );
};

const mapStateToProps = state => ({
  error: state.authLogin.error,
  user: state.authLogin.user,
  isAuthenticated: state.authLogin.isAuthenticated,
  isSubmit: state.authLogin.isSubmit,
  isCompleted: state.authLogin.isCompleted
});

/* istanbul ignore next */
export const onSignIn = user => signInAction(user);
export const cleanUp = () => cleanSignIn();

export const SignInComponent = SignIn;

export default connect(mapStateToProps, { onSignIn, cleanUp })(SignIn);