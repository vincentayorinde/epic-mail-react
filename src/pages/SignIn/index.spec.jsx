import React from 'react';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { BrowserRouter as Router } from 'react-router-dom';
import SignIn, { SignInComponent, cleanUp } from './index';
import store from '../../redux/storeConfig';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Sign In View Component', () => {
  it('Should render without errors', () => {
    const component = shallow(
      <Router>
        <SignIn store={store} />
      </Router>
    );
    expect(component).toMatchSnapshot();
  });

  it('Should Render With Children Component', () => {
    const component = mount(
      <Router>
        <SignIn store={store} />
      </Router>
    );
    expect(component).toMatchSnapshot();
    expect(component.find('button')).toHaveLength(1);
    expect(component.find('input')).toHaveLength(2);
  });

  it('Should intiate onChange Props', () => {
    const event = {
      preventDefault() { },
      target: { value: 'vincent', name: 'email' }
    };
    const mockOnSignInFn = jest.fn();
    const component = mount(
      <Router>
        <SignInComponent onSignIn={mockOnSignInFn} />
      </Router>
    );
    const inputTag = component.find('input').at(0);
    const submitBtn = component.find('button').at(0);
    inputTag.simulate('change', event);
    submitBtn.simulate('submit',
      { preventDefault() { } });
    expect(mockOnSignInFn.mock.calls.length).toBe(1);
  });

  test('USE EFFECT => Should set toast if isCompleted is true', () => {
    const signInStore = mockStore({
      authLogin:
      {
        error: null,
        user: {},
        isAuthenticated: false,
        isCompleted: true,
        isSubmit: false
      }
    });
    const component = mount(
      <Router>
        <SignIn store={signInStore} />
      </Router>
    );

    const p = component.find('p').at(5);
    expect(p.text()).toEqual('Email');
  });

  test('Should initiate errors if error exist and is a string', () => {
    const signInStore = mockStore({
      authLogin:
      {
        error: 'errors',
        user: {},
        isAuthenticated: false,
        isCompleted: false,
        isSubmit: false
      }
    });
    const component = mount(
      <Router>
        <SignIn store={signInStore} />
      </Router>
    );

    const p = component.find('p').at(5);
    expect(p.text()).toEqual('Email');
  });

  test('Should initiate errors if error exist and is an array', () => {
    const signInStore = mockStore({
      authLogin:
      {
        error: [{ message: 'errors' }],
        user: {},
        isAuthenticated: false,
        isCompleted: false,
        isSubmit: false
      }
    });
    const component = mount(
      <Router>
        <SignIn store={signInStore} />
      </Router>
    );

    const p = component.find('p').at(5);
    expect(p.text()).toEqual('Email');
  });

  test('Should run clean up action ', () => {
    expect(cleanUp()).toEqual({ type: 'CLEAN_SIGNIN' });
  });

});