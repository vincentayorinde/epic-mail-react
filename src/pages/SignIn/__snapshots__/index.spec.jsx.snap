// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign In View Component Should Render With Children Component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(SignIn)
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <SignIn
        cleanUp={[Function]}
        error={null}
        isAuthenticated={false}
        isCompleted={false}
        isSubmit={false}
        onSignIn={[Function]}
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        user={Object {}}
      >
        <div
          className="container-1"
        >
          <div
            className="box-1"
          >
            <div
              className="content"
            >
              <p>
                Experience
              </p>
              <p>
                the new way to
              </p>
              <p>
                 send messages
              </p>
              <p>
                Use EPIC Mail 
                <em>
                  today!
                </em>
              </p>
              <div
                className="links"
              >
                <ul>
                  <li>
                    <a
                      href="#"
                    >
                      Terms
                    </a>
                  </li>
                  <li>
                    <a
                      href="#"
                    >
                      Privacy
                    </a>
                  </li>
                  <li>
                    <a
                      href="#"
                    >
                      APIs
                    </a>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <div
            className="box-2"
          >
            <div
              className="signup"
            >
              <p>
                Welcome to EPIC Mail
              </p>
              <h2>
                Sign In
              </h2>
              <form
                onSubmit={[Function]}
              >
                <p>
                  Email
                </p>
                <Input
                  id="email"
                  inputType="email"
                  name="email"
                  onChange={[Function]}
                  pattern="^[\\\\w.]+@[\\\\w]{2,20}.[a-z]{2,10}$"
                  required={true}
                  title="user@email.com"
                >
                  <input
                    id="email"
                    name="email"
                    onChange={[Function]}
                    pattern="^[\\\\w.]+@[\\\\w]{2,20}.[a-z]{2,10}$"
                    required={true}
                    title="user@email.com"
                    type="email"
                  />
                </Input>
                <p>
                  Password
                </p>
                <Input
                  id="password"
                  name="password"
                  onChange={[Function]}
                  required={true}
                  type="password"
                >
                  <input
                    id="password"
                    name="password"
                    onChange={[Function]}
                    required={true}
                    type="password"
                  />
                </Input>
                <Button
                  isSubmit={false}
                  name="Sign In"
                  type="submit"
                >
                  <button
                    disabled={false}
                    type="submit"
                  >
                    Sign In
                  </button>
                </Button>
              </form>
              <br />
              <br />
              <i
                className="fas fa-chevron-right"
              />
               Don't have an account? 
              <Link
                to="/"
              >
                <a
                  href="/"
                  onClick={[Function]}
                >
                  Sign Up
                </a>
              </Link>
            </div>
          </div>
          <div
            className="box-3"
          >
            <div
              className="content"
            />
          </div>
        </div>
      </SignIn>
    </Connect(SignIn)>
  </Router>
</BrowserRouter>
`;

exports[`Sign In View Component Should render without errors 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(SignIn)
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />
</Router>
`;
